pkg/liquidity-mining/contracts/fee-distribution/FeeDistributor.sol

```diff
@@ -0,0 +1,410 @@
+        require(block.timestamp > _startTime, "Fee distribution has not started yet");
+        
+        uint256 tokensLength = tokens.length;
+        for(uint256 i = 0; i < tokensLength; ++i) {
+            _checkpointToken(tokens[i], true);
+        }
+    }
+
+    function claimToken(address user, IERC20 token) external nonReentrant returns (uint256) {
+        // Prevent someone from assigning tokens to an inaccessible week.
+        require(block.timestamp > _startTime, "Fee distribution has not started yet");
+        _checkpointTotalSupply();
+        _checkpointToken(token, false);
+        _checkpointUserBalance(user);
+
+        uint256 amount = _claimToken(user, token);
+        return amount;
+    }
+
+    function claimTokens(address user, IERC20[] calldata tokens) external nonReentrant returns (uint256[] memory) {
+        // Prevent someone from assigning tokens to an inaccessible week.
+        require(block.timestamp > _startTime, "Fee distribution has not started yet");
+        _checkpointTotalSupply();
+        _checkpointUserBalance(user);
+
+        uint256 tokensLength = tokens.length;
+        uint256[] memory amounts = new uint256[](tokensLength);
+        for(uint256 i = 0; i < tokensLength; ++i) {
+            _checkpointToken(tokens[i], false);
+            amounts[i] = _claimToken(user, tokens[i]);
+        }
+
+        return amounts;
+    }
+    
+
+    // Internal functions
+
+    /**
+     * @dev It is required that both the global, token and user state have been properly checkpointed
+     * before calling this function.
+     */
+    function _claimToken(address user, IERC20 token) internal returns (uint256) {
+        uint256 userTimeCursor = _getUserTokenTimeCursor(user, token);
+        // We round `_tokenTimeCursor` down so it represents the beginning of the first incomplete week.
+        uint256 currentActiveWeek = _roundDownTimestamp(_tokenTimeCursor[token]);
+        mapping(uint256 => uint256) storage tokensPerWeek = _tokensPerWeek[token];
+        mapping(uint256 => uint256) storage userBalanceAtTimestamp = _userBalanceAtTimestamp[user];
+
+        uint256 amount;
+        for (uint256 i = 0; i < 20; ++i) {
+            // We only want to claim for complete weeks so break once we reach `currentActiveWeek`.
+            // This is as `tokensPerWeek[currentActiveWeek]` will continue to grow over the week.
+            if (userTimeCursor >= currentActiveWeek) break;
+
+            amount +=
+                (tokensPerWeek[userTimeCursor] * userBalanceAtTimestamp[userTimeCursor]) /
+                _veSupplyCache[userTimeCursor];
+            userTimeCursor += 1 weeks;
+        }
+        _userTokenTimeCursor[user][token] = userTimeCursor;
+
+        if (amount > 0) {
+            _tokenLastBalance[token] -= amount;
+            token.safeTransfer(user, amount);
+            emit TokensClaimed(user, token, amount, userTimeCursor);
+        }
+
+        return amount;
+    }
```

Comment:
As a knock on consequence we need to add this function to handle claiming of tokens with each token getting its own time cursor to ensure that a user can only claim fees from a given week once.